Nutriplan

purplebird opened the pull request:
I don't know what I'm doing.

Possibly fixed #210???

rolandgeider commented:
Nobody does, @purplebird, nobody does ;)

rolandgeider commented on wger/nutrition/models.py:
```diff
@@ -137,10 +138,17 @@ def get_nutritional_values(self):
+                if not closest_entry or \
+                (abs(entry.date - self.creation_date) <= abs(closest_entry.date - self.creation_date)):
+                    closest_entry = entry
+            weight = Decimal(closest_entry.weight)
```
Better than iterating through all the entries, I'd do something [similar to this](https://stackoverflow.com/questions/15855715/filter-on-datetime-closest-to-the-given-datetime). I also think it's possible to replace all the try-excepts by using ``[....].order_by('date').first()``, but I'm not 100% sure

purplebird replied:
Ironically, I actually had it that way originally.

The try-excepts should be unnecessary since we're using less than or equal to and have already confirmed one entry exists with the conditional on 143. Please correct me if you think my logic is off.

rolandgeider replied:
You are right, there will be at least one weight entry when we reach this code.

rolandgeider commented:
Oh, don't forget to add yourself to the AUTHORS file

purplebird commented:
Done!

rolandgeider commented:
I have now taken a look at the errors from travis. Doing a ``.order_by('date')[0]`` doesn't work when the user has no weight entries at all. We'll have to use [first()](https://docs.djangoproject.com/en/1.8/ref/models/querysets/#first), which simply returns None if there are no entries.

As a nice side effect, the tests also check that the correct weight entry is selected in one of the calculations, so we'd just have to update the assertEqual there. The PEP8 formatting errors are easy to fix once you know waht to look for, you can just do a ``pep8 wger`` in the source to check for this.

purplebird commented:
@rolandgeider , I made a couple sweeping changes here.

I moved the closest entry code into its own function for reusability. I changed the get_nutritional_data() function to now take a weight entry to base the per body weight values. I updated the view to use the new get_closest_weight_entry function for this. I also changed the template so now the user can see which entry the values are based on.

I'm still learning how Django and the template system works, so please forgive small errors. :)

rolandgeider commented on wger/nutrition/templates/plan/view.html:
```diff
@@ -186,6 +186,9 @@
 
 
 <h4>{% trans "Nutritional data" %}</h4>
+{% if closest_weight_entry != None %}
```
You can just do ``{% if closest_weight_entry %}`` like in regular python

rolandgeider commented on wger/nutrition/templates/plan/view.html:
```diff
@@ -186,6 +186,9 @@
 
 <h4>{% trans "Nutritional data" %}</h4>
+{% if closest_weight_entry != None %}
+    <h5>{% trans "Based on the weight entry dated" %} {{closest_weight_entry.date}} ({{closest_weight_entry.weight}})</h5>
```
Better do here

```
{% blocktrans with date=closest_weight_entry.date  weight=... %}
Based on the weight entry dated {{date}} ({{weight}})
{% endblocktrans %}
```

Then the context is clearer when translating the string

purplebird replied:
Thanks.

This looks a lot cleaner.

rolandgeider commented:
Looking good!

BTW, you can run the test locally, just do a ``python manage.py test`` to run the full test suite (you can also do a ``manage.py test wger.nutrition`` to only run the nutrition tests or even ``manage.py test wger.nutrition.test_plan``)

purplebird commented:
Yeah, I managed to figure that out near the end of my hacking. ;)

rolandgeider commented:
I've thought a bit about this. I think it would be better if ``get_nutritional_values`` would use the closest weight by default, since this is the more "correct" behaviour most of the time. Perhaps something like this? I like the idea of using this to compare differences over time but for the moment the two results we can get ("with weight from back then" and "with current weight") are probably enough. We can expand that in another issue, though. Does this sound OK to you @purplebird ?

```python
def get_nutritional_values(self, use_closest_weight=True):
    if use_closest_weight:
        weight = self.get_closest_weight_entry().weight
    else:
        weight = self.user.userprofile.weight
```

purplebird commented:
We could do it this way. There is one disadvantage: you would have to ping the database twice to get the weight info for the template (since you're no longer passing it into get_nutritional_values(). Granted, this is relatively minor... Up to you!

rolandgeider commented:
You are right, but I think we can live with that. On the other hand, we could perhaps return the weight used in the per_kg dictionary. Something like this:
```python
    'per_kg': {'protein': 123,
               'carbohydrates': 123,
               'fat': 123,
               'weight_used': {
                   'weight': 123,
                   'date': 123
               }
    }
```

purplebird commented:
That is what I originally did, but it explodes where you hit this line:

        # Only 2 decimal places, anything else doesn't make sense
        for key in result.keys():
            for i in result[key]:
                result[key][i] = Decimal(result[key][i]).quantize(TWOPLACES)

And the answer seemed to be either explicitly ignore the 'weight_used' category or do something else here.

rolandgeider commented:
You are right, lets just fire another query in the view

rolandgeider commented:
How is it going @purplebird? If you could implement the changes I would merge the PR (I know I have been nitpicking a lot here, but this is the latest change, I promise!)

purplebird commented:
Sorry!!! I've been distracted working on AI stuff. :)

On Mon, Nov 30, 2015 at 1:17 PM, Roland Geider <notifications@github.com>
wrote:

> How is it going @purplebird <https://github.com/purplebird>? If you could
> implement the changes I would merge the PR (I know I have been nitpicking a
> lot here, but this is the latest change, I promise!)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rolandgeider/wger/pull/231#issuecomment-160711836>.
>

rolandgeider commented:
That sounds so cool! Reading and doing some machine learning is also in my personal todo list

purplebird commented:
Okay. I think that should do it. :)

rolandgeider commented:
Merged. Thanks again!
